### PR TITLE
#14 Add option -f to CLI to modify the original app folder

### PR DESCRIPTION
**Pull request by mmcrockett**
@fabiocarvalho777

### CODE REVIEW

**Comment by fabiocarvalho777** [comment]
@mmcrockett the PR title is correct, but please rename the commit message. After this is merged, what stays in the history is really the commit message, not the PR title.

File: butterfly-cli/src/main/java/com/paypal/butterfly/cli/ButterflyCliOption.java [hunk 1]
```diff
@@ -27,6 +27,7 @@
     protected static final String CLI_OPTION_TEMPLATE_SHORTCUT = "s";
     protected static final String CLI_OPTION_UPGRADE_VERSION = "u";
     protected static final String CLI_OPTION_RESULT_FILE = "r";
+    protected static final String CLI_OPTION_IN_PLACE_TRANSFORM = "f";
```
Comment (fabiocarvalho777) [inline butterfly-cli/src/main/java/com/paypal/butterfly/cli/ButterflyCliOption.java]: Please, rename this to `CLI_OPTION_MODIFY_ORIGINAL_FOLDER`. That is better aligned with `-f`.

File: butterfly-cli/src/main/java/com/paypal/butterfly/cli/ButterflyCliOption.java [hunk 2]
```diff
@@ -77,6 +78,9 @@
                 .withRequiredArg()
                 .ofType(File.class)
                 .describedAs("result file");
+
+        // In place option
```
Comment (fabiocarvalho777) [inline butterfly-cli/src/main/java/com/paypal/butterfly/cli/ButterflyCliOption.java]: Please, change the comment to `// Modify original folder option`

File: butterfly-cli/src/main/java/com/paypal/butterfly/cli/ButterflyCliOption.java [hunk 3]
```diff
@@ -77,6 +78,9 @@
...
+        optionParser.accepts(CLI_OPTION_IN_PLACE_TRANSFORM, "Transforms the application in the same folder as the original content, cannot be used with output folder option.");
```
Comment (fabiocarvalho777) [inline butterfly-cli/src/main/java/com/paypal/butterfly/cli/ButterflyCliOption.java]: Actually, if `-f` is used, then `-o` and `-z` are simply ignored (as specified in #14 ). Please change the message to: `"Transforms the application in the same folder as the original content. Options (-o) or (-z) are ignored if (-f) is specified"`

File: butterfly-cli/src/main/java/com/paypal/butterfly/cli/ButterflyCliRunner.java [hunk 4]
```diff
@@ -99,6 +99,19 @@ public ButterflyCliRun run() throws IOException {
         run.setApplication(applicationFolder);
 
         File transformedApplicationFolder = (File) optionSet.valueOf(CLI_OPTION_TRANSFORMED_APP_FOLDER);
+
+        if (optionSet.has(CLI_OPTION_IN_PLACE_TRANSFORM) && optionSet.has(CLI_OPTION_TRANSFORMED_APP_FOLDER)) {
```
Comment (fabiocarvalho777) [inline butterfly-cli/src/main/java/com/paypal/butterfly/cli/ButterflyCliRunner.java]: No need to check, `-z` and `-o` are ignored if `-f` is specified.

File: butterfly-cli/src/main/java/com/paypal/butterfly/cli/ButterflyCliRunner.java [hunk 5]
```diff
@@ -99,6 +99,19 @@ public ButterflyCliRun run() throws IOException {
...
+            String errorMessage = String.format("Can't specify output folder (%s) AND in place (%s) option at same time.", CLI_OPTION_TRANSFORMED_APP_FOLDER, CLI_OPTION_IN_PLACE_TRANSFORM);
+            registerError(run, errorMessage);
+            return run;
+        }
+
+        if (optionSet.has(CLI_OPTION_IN_PLACE_TRANSFORM) && optionSet.has(CLI_OPTION_CREATE_ZIP)) {
```
Comment (fabiocarvalho777) [inline butterfly-cli/src/main/java/com/paypal/butterfly/cli/ButterflyCliRunner.java]: No need to check, -z and -o are ignored if -f is specified.

File: butterfly-cli/src/main/java/com/paypal/butterfly/cli/ButterflyCliRunner.java [hunk 6]
```diff
@@ -132,7 +145,7 @@ public ButterflyCliRun run() throws IOException {
             logger.info("-z option has been set, transformed application will be placed into a zip file");
         }
 
-        Configuration configuration = new Configuration(transformedApplicationFolder, createZip);
+        Configuration configuration = new Configuration(transformedApplicationFolder, createZip, optionSet.has(CLI_OPTION_IN_PLACE_TRANSFORM));
```
Comment (fabiocarvalho777) [inline butterfly-cli/src/main/java/com/paypal/butterfly/cli/ButterflyCliRunner.java]: If `-f` is specified, then you can ignore the `transformedApplicationFolder` and `createZip`, and just use `Configuration` default constructor instead.

File: butterfly-facade/src/main/java/com/paypal/butterfly/facade/Configuration.java [hunk 7]
```diff
@@ -14,6 +14,7 @@
     // See the setters for information about each property
     private File outputFolder = null;
     private boolean zipOutput = false;
+    private boolean modifyInPlace = false;
```
Comment (fabiocarvalho777) [inline butterfly-facade/src/main/java/com/paypal/butterfly/facade/Configuration.java]: Please, rename this to `modifyOriginalFolder`, then rename setter and getter accordingly.

File: butterfly-facade/src/main/java/com/paypal/butterfly/facade/Configuration.java [hunk 8]
```diff
@@ -36,6 +37,20 @@ public Configuration(File outputFolder, boolean zipOutput) {
...
 
+    /**
+     * Butterfly default configuration
+     *
+     * @param outputFolder the output folder where the transformed application is
+     *                     supposed to be placed
+     * @param zipOutput if true, the transformed application folder will be compressed into a zip file
+     * @param modifyInPlace if true, the transformation will occur in application folder
+     */
+    public Configuration(File outputFolder, boolean zipOutput, boolean modifyInPlace) {
```
Comment (fabiocarvalho777) [inline butterfly-facade/src/main/java/com/paypal/butterfly/facade/Configuration.java]: No need to add a new constructor. If `-f` is present, then the default public constructor can be used instead.

**Comment by mmcrockett** [comment]
@fabiocarvalho777 made your changes

**Comment by fabiocarvalho777** [comment]
@mmcrockett thanks a lot for the changes!!

In terms of commit message, I see you changed it, but you missed the hashtag before the issue id. You might have done it in vi, and had to remove it to avoid it being interpreted as comment. If that is what happened, you can avoid that by adding a space before the hashtag. Let me know if this doesn't make sense.

File: butterfly-facade/src/test/java/com/paypal/butterfly/facade/ConfigurationTest.java [hunk 9]
```diff
@@ -17,15 +17,19 @@
     @Test
     public void testEqualsAndHashCode(){
         //Equals use case
-        Configuration configuration = new Configuration();
+        Configuration configuration       = new Configuration();
```
Comment (fabiocarvalho777) [inline butterfly-facade/src/test/java/com/paypal/butterfly/facade/ConfigurationTest.java]: Please remove these spaces before `=`

File: butterfly-facade/src/test/java/com/paypal/butterfly/facade/ConfigurationTest.java [hunk 10]
```diff
@@ -59,6 +63,16 @@ public void testZipOutput(){
         Assert.assertTrue(configuration.isZipOutput());
     }
 
+    /**
```
Comment (fabiocarvalho777) [inline butterfly-facade/src/test/java/com/paypal/butterfly/facade/ConfigurationTest.java]: Please, also add a test making sure that, if the constructor with parameters is used, then `isModifyOriginalFolder` returns `false`.

**Comment by mmcrockett** [comment]
@fabiocarvalho777 I can't add the test, I tried but it modifies the directory (like it's supposed to) which causes the other tests to fail because it's changed the sample-app...if that makes sense.

**Comment by fabiocarvalho777** [comment]
@mmcrockett you could add a method to ButterflyIT annotated with TestNG `@BeforeClass` where you would duplicate `sample-app`, placing the duplication under the test `target` folder, then you could set a instance variable pointing to that, and then you can add a test using the duplicated app folder.

**Comment by mmcrockett** [comment]
@fabiocarvalho777 Added integration test.

**Comment by fabiocarvalho777** [comment]
@mmcrockett for some reason (not related to your changes) integration tests didn't get run in the ECI build ([as seen here](https://travis-ci.org/paypal/butterfly/builds/355944784?utm_source=github_status&utm_medium=notification)). Would you find out why? Thanks.

**Comment by mmcrockett** [comment]
@fabiocarvalho777 Are you sure they didn't run? I see my '-f' test ran since there is a line in the log 'Transformed application folder: /home/travis/build/paypal/butterfly/tests/integration-tests/./target/sample-app-copy' - check the full raw log.

**Comment by fabiocarvalho777** [comment]
@mmcrockett you are right, it did run. I missed it, sorry. I am going to merge this PR now. Thanks a lot for being patient Mike!

**Pull request merged**